openedx/core/djangoapps/lang_pref/middleware.py

```diff
@@ -21,17 +26,46 @@ class LanguagePreferenceMiddleware(object):
     def process_request(self, request):
         """
         If a user's UserPreference contains a language preference, use the user's preference.
+        Save the current language preference cookie as the user's preferred language.
         """
-        languages = released_languages()
-        system_released_languages = [seq[0] for seq in languages]
+        cookie_lang = request.COOKIES.get(settings.LANGUAGE_COOKIE, None)
+        if cookie_lang:
+            if request.user.is_authenticated():
+                set_user_preference(request.user, LANGUAGE_KEY, cookie_lang)
```

Comment:
? We're setting the cookie based on the user preference (below), and then overriding the user preference with the value from the cookie?

1. Is there an issue with trusting the value in the cookie over what the user explicitly put in their preferences?
2. Why doesn't this add a db write to every request?  (I saw the tests.  All the query counts are going down, so I'm obviously wrong here. Help me see why.)